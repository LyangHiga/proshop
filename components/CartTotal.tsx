import Link from "next/link";
import { useSelector } from "react-redux";
import {
  Card,
  CardContent,
  CardActions,
  Button,
  Divider,
  Grid,
  Typography,
} from "@material-ui/core";

import { RootState } from "../store/reducers/reducers";
import Item from "../models/Item";

import useStyles from "../styles/CartTotalStyles";

const CartTotal = () => {
  const classes = useStyles();
  const cartItems = useSelector(
    (state: RootState) => state.cart.cartItems
  ) as Item[];
  return (
    <Card className={classes.card}>
      <CardContent className={classes.cardContent}>
        <Grid container className={classes.cardTextContainer}>
          <Typography variant="overline" className={classes.cardTitle}>
            Subtotal ({cartItems.reduce((acc, item) => acc + item.quantity, 0)})
            Items:
          </Typography>
        </Grid>
        <Grid
          container
          alignItems="center"
          justify="center"
          className={classes.cardTextContainer}
        >
          <Typography variant="overline" className={classes.cardText}>
            $
            {(
              Math.round(
                (cartItems.reduce(
                  (acc, item) => acc + item.quantity * item.price,
                  0
                ) +
                  Number.EPSILON) *
                  100
              ) / 100
            ).toLocaleString("pt-br", { maximumFractionDigits: 2 })}
          </Typography>
        </Grid>
        <Divider className={classes.divider} />
      </CardContent>
      <CardActions>
        <Grid container justify="center" alignItems="center">
          <Grid item>
            <Link href="/shipping">
              <Button
                name="btn-card-add-to-cart"
                variant="contained"
                color="primary"
                size="large"
              >
                Proceed to Checkout
              </Button>
            </Link>
          </Grid>
        </Grid>
      </CardActions>
    </Card>
  );
};

export default CartTotal;
