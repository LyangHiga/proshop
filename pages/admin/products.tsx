import { GetServerSideProps } from "next";
import { Grid, Typography } from "@material-ui/core";

import Product from "../../models/Product";

import ProductListItem from "../../components/ProductListItem";
import Header from "../../components/Header";
import Footer from "../../components/Footer";

import useStyles from "../../styles/adminUsersStyles";

interface ProductsProps {
  products: Product[];
  token: string;
}

// TODO: Token will be used to make http req in ProductListItem
const products = ({ products, token }: ProductsProps) => {
  const classes = useStyles();
  return (
    <div>
      <Header />
      <main className="main">
        <Typography variant="h4" className={classes.text}>
          Products
        </Typography>
        <Grid container>
          <Grid item container className={classes.container}>
            <Grid item md={3}>
              <Typography variant="h6" className={classes.text}>
                ID
              </Typography>
            </Grid>
            <Grid item md={3}>
              <Typography variant="h6" className={classes.text}>
                Name
              </Typography>
            </Grid>
            <Grid item md={1}>
              <Typography variant="h6" className={classes.text}>
                Price
              </Typography>
            </Grid>
            <Grid item md={1}>
              <Typography variant="h6" className={classes.text}>
                Category
              </Typography>
            </Grid>
            <Grid item md={2}>
              <Typography variant="h6" className={classes.text}>
                Brand
              </Typography>
            </Grid>
            <Grid item md={1}>
              <Typography variant="h6" className={classes.text}>
                Remove
              </Typography>
            </Grid>
            {products.map((product) => (
              <ProductListItem product={product} key={product._id} />
            ))}
          </Grid>
        </Grid>
      </main>
      <Footer />
    </div>
  );
};

export const getServerSideProps: GetServerSideProps = async ({ req }) => {
  // to check login and isAdmin
  const { user } = req.cookies;
  if (!user) {
    return {
      redirect: {
        destination: "/login",
        permanent: false,
      },
    };
  }
  const { token, isAdmin } = JSON.parse(user);
  if (!token || !isAdmin) {
    return {
      redirect: {
        destination: "/login",
        permanent: false,
      },
    };
  }

  // To get products list
  const res = await fetch("http://localhost:5000/api/products", {
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  });

  // Dont return if status is not 200
  // products array will be empty, and page return null
  // if dont retun null will broke because products will be an empty array
  if (res.status !== 200 || !res.ok) {
    return {
      redirect: {
        destination: "/login",
        permanent: false,
      },
    };
  }

  const products = (await res.json()) as Product[];

  return {
    props: {
      products,
      token,
    },
  };
};

export default products;
