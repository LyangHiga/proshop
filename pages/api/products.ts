// JUST TO TEST GetStaticProps
import products from "../../products";
// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import { NextApiRequest, NextApiResponse } from "next";

export default (req: NextApiRequest, res: NextApiResponse) => {
  res.setHeader("Cache-Control", "s-maxage=60, stale-while-revalidate");

  res.status(200).json({ products });
};
