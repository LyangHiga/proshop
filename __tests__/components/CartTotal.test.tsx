import { shallow } from "enzyme";
import * as Redux from "react-redux";

import CartTotal from "../../components/CartTotal";
import ItemModel from "../../models/Item";
import products from "../../products";

describe("Cart Total", () => {
  let useSelectorSpy: jest.SpyInstance;

  beforeEach(() => {
    useSelectorSpy = jest.spyOn(Redux, "useSelector");
  });

  test("renders correctly", () => {
    const items = new Array<ItemModel>();
    for (let p of products) {
      const item = { ...p, quantity: 6 };
      items.push(item);
    }

    useSelectorSpy.mockReturnValue(items);
    const wrapper = shallow(<CartTotal />);

    expect(wrapper).toMatchSnapshot();
  });
});
