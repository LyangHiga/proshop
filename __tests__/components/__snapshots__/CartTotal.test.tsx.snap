// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Total renders correctly 1`] = `ShallowWrapper {}`;
